File: BAEKJOON/17073 - 나무 위의 빗물/나무 위의 빗물.py
import sys
from collections import defaultdict

def main():
    # 입력을 받기 위해 sys.stdin.read를 사용하여 전체 데이터를 읽음
    input = sys.stdin.read
    # 입력받은 데이터를 줄 단위로 나누어 리스트로 변환
    data = input().splitlines()

    # 첫 번째 줄에서 노드의 수 N과 1번 노드에 고인 물의 양 W를 추출
    N, W = map(int, data[0].split())
    
    # 트리의 연결 관계를 저장할 인접 리스트 형태의 딕셔너리를 생성
    tree = defaultdict(list)
    
    # 두 번째 줄부터 N-1줄까지의 간선 정보를 읽어와 트리 구성
    for line in data[1:]:
        U, V = map(int, line.split())
        tree[U].append(V)  # U 노드에 V 노드를 연결
        tree[V].append(U)  # V 노드에 U 노드를 연결
    
    # 리프 노드(자식이 없는 노드)의 개수를 셀 변수
    leaf_count = 0
    # 2번 노드부터 N번 노드까지 순회하여 리프 노드 찾기
    for node in range(2, N+1):
        # 각 노드의 연결된 노드 개수가 1개인 경우, 해당 노드는 리프 노드
        if len(tree[node]) == 1:
            leaf_count += 1
    
    # 리프 노드의 개수로 1번 노드의 물을 나누어 각 리프 노드에 고이는 물의 양을 계산
    result = W / leaf_count
    # 결과를 소수점 10자리까지 출력
    print(f"{result:.10f}")

if __name__ == "__main__":
    # main 함수 호출
    main()

File: BAEKJOON/17073 - 나무 위의 빗물/나무 위의 빗물.cs
using System;
using System.Collections.Generic;

class Program
{
    static void Main()
    {
        // 첫 번째 입력: 노드 수와 물의 양을 저장할 변수
        string[] input = Console.ReadLine().Split();
        int N = int.Parse(input[0]); // 노드 수
        double W = double.Parse(input[1]); // 1번 노드에 고인 물의 양
        
        // 트리의 인접 리스트를 저장할 딕셔너리
        Dictionary<int, List<int>> tree = new Dictionary<int, List<int>>();
        
        // 각 노드에 연결된 간선 입력 받기
        for (int i = 0; i < N - 1; i++)
        {
            input = Console.ReadLine().Split();
            int U = int.Parse(input[0]);
            int V = int.Parse(input[1]);
            
            if (!tree.ContainsKey(U))
                tree[U] = new List<int>(); // U에 리스트 생성
            if (!tree.ContainsKey(V))
                tree[V] = new List<int>(); // V에 리스트 생성
            
            tree[U].Add(V); // U와 V 연결
            tree[V].Add(U); // V와 U 연결
        }
        
        // 리프 노드의 개수를 저장할 변수
        int leafCount = 0;
        
        // 2번 노드부터 N번 노드까지 탐색하여 리프 노드 확인
        for (int node = 2; node <= N; node++)
        {
            if (tree[node].Count == 1) // 연결된 노드가 하나인 경우 리프 노드
                leafCount++;
        }
        
        // 리프 노드에 고인 물의 양 계산
        double result = W / leafCount;
        
        // 결과를 소수점 10자리까지 출력
        Console.WriteLine($"{result:F10}");
    }
}
